crates/extensions/c8y_mapper_ext/src/operations/software_update.rs

```diff
@@ -1,79 +1,73 @@
 use c8y_api::smartrest;
 use c8y_api::smartrest::smartrest_serializer::CumulocitySupportedOperations;
 use tedge_api::mqtt_topics::EntityTopicId;
-use tedge_api::workflow::GenericCommandState;
 use tedge_api::CommandStatus;
 use tedge_api::SoftwareListCommand;
 use tedge_api::SoftwareUpdateCommand;
 use tedge_mqtt_ext::MqttMessage;
 
-use crate::error::ConversionError;
-
+use super::error::OperationError;
 use super::EntityTarget;
 use super::OperationContext;
+use super::OperationResult;
 
 impl OperationContext {
     pub async fn publish_software_update_status(
         &self,
-        target: EntityTarget,
+        target: &EntityTarget,
         cmd_id: &str,
         message: &MqttMessage,
-    ) -> Result<(Vec<MqttMessage>, Option<GenericCommandState>), ConversionError> {
+    ) -> Result<OperationResult, OperationError> {
         let command = match SoftwareUpdateCommand::try_from_bytes(
             target.topic_id.clone(),
             cmd_id.to_string(),
             message.payload_bytes(),
-        )? {
+        )
+        .context("Could not parse command as a software update command")?
+        {
             Some(command) => command,
             None => {
                 // The command has been fully processed
-                return Ok((vec![], None));
+                return Ok(OperationResult::Ignored);
             }
         };
 
-        let topic = target.smartrest_publish_topic;
-        let messages = match command.status() {
+        let topic = &target.smartrest_publish_topic;
+        match command.status() {
             CommandStatus::Init | CommandStatus::Scheduled | CommandStatus::Unknown => {
                 // The command has not been processed yet
-                vec![]
-            }
-            CommandStatus::Executing => {
-                let smartrest_set_operation_status =
-                    smartrest::smartrest_serializer::set_operation_executing(
-                        CumulocitySupportedOperations::C8ySoftwareUpdate,
-                    );
-                vec![MqttMessage::new(&topic, smartrest_set_operation_status)]
+                Ok(OperationResult::Ignored)
             }
+            CommandStatus::Executing => Ok(OperationResult::Executing),
             CommandStatus::Successful => {
                 let smartrest_set_operation =
                     smartrest::smartrest_serializer::succeed_operation_no_payload(
                         CumulocitySupportedOperations::C8ySoftwareUpdate,
                     );
 
-                vec![
-                    MqttMessage::new(&topic, smartrest_set_operation),
-                    command.clearing_message(&self.mqtt_schema),
-                    self.request_software_list(&target.topic_id),
-                ]
+                Ok(OperationResult::Finished {
+                    messages: vec![
+                        MqttMessage::new(topic, smartrest_set_operation),
+                        self.request_software_list(&target.topic_id),
```

Comment:
Thanks to this refactoring PR, there is now an opportunity to remove from here the software list request, as command finalization in now done in one place

Unfortunately, this is not enough to remove the need for an id generator clone (currently both the converter and the running operation for the software update need to generate id). An option could be to delegate the `OperationHandler`  the generation of operation ids.

Reply:
> Thanks to this refactoring PR, there is now an opportunity to remove from here the software list request, as command finalization in now done in one place

I'm not sure if removing `software_list` publish from `software_update` is practical:

1. Local `software_update` operation is published on local MQTT broker in `c8y_mapper_ext/src/converter.rs:770`, in function `forward_software_request`
2. This local `software_update` operation is picked up by the `OperationHandler` which runs `c8y_mapper_ext::operations::software_update::OperationContext::publish_software_update_status`
3. `publish_software_update_status` returns "succeed operation" message with no payload, and invokes a local `software_list` operation after `software_update` operation is succeeded or failed.
4. `publish_software_list`, when it's successful, sends the software list via HTTP proxy or smartrest.

What I'm seeing is that `publish_software_list` isn't an operation, and it can be invoked anytime, just after `software_update` software list changes, so it means it has to be updated. `software_list` isn't necessary to finalize the `software_update` operation (we send together success/failure of `software_update` and invocation of `software_list`), and `software_update` is the only operation that does it, so it makes little sense to extract it.

The only thing that's problematic is sharing this id generator, because generating the ids is necessary for creating local operations. Right now, creating local operations is also spread among methods in `CumulocityConverter`: `forward_software_request`, `forward_restart_request`, `forward_operation_request`, but also `convert_log_upload_request`, `convert_config_snapshot_request` etc. in `operations` module. They all do the same thing (converting c8y operation messages to local thin-edge operation messages), so they should be put together in a single place, either in `OperationHandler`, or in some other object that will have smaller scope. It would then contain the `IdGenerator` and provide methods to start new local operations by publishing on the local operation topics. `software_update` would then simply call this to create a new `software_list` operation instead of doing it manually.

Reply:
> What I'm seeing is that `publish_software_list` isn't an operation, and it can be invoked anytime, just after `software_update` software list changes, so it means it has to be updated. `software_list` isn't necessary to finalize the `software_update` operation (we send together success/failure of `software_update` and invocation of `software_list`), and `software_update` is the only operation that does it, so it makes little sense to extract it.

Yes, sending a software list after a software update is not strictly required by c8y and is always done at the device initiative. This is done in two cases: when the  software update capability is added and after each software update request.

Technically, I see no reason to have this done inside the software update handler or elsewhere.
My concern is that one should have *only* one component that trigger operations on behalf of c8y.

> The only thing that's problematic is sharing this id generator, because generating the ids is necessary for creating local operations.

We are fully aligned here.

> Right now, creating local operations is also spread among methods in `CumulocityConverter`: `forward_software_request`, `forward_restart_request`, `forward_operation_request`, but also `convert_log_upload_request`, `convert_config_snapshot_request` etc. in `operations` module. They all do the same thing (converting c8y operation messages to local thin-edge operation messages), so they should be put together in a single place, either in `OperationHandler`, or in some other object that will have smaller scope. It would then contain the `IdGenerator` and provide methods to start new local operations by publishing on the local operation topics. `software_update` would then simply call this to create a new `software_list` operation instead of doing it manually.

As you say: *they should be put together in a single place, either in `OperationHandler` or in some other object.*
I'm looking for something symmetric:
- the object creating an operation request should also be the object that clears that operation request when done
- the `OperationHandler` seems to be a nice candidate for that
- when clearing a software update this object could also possibly trigger a following software list request 
- in case you wonder: I'm not requesting to do that in this PR.